src/Nethermind/Nethermind.Trie/TrieNode.Visitor.cs

```diff
@@ -187,31 +190,43 @@ void VisitSingleThread(ref TreePath parentPath, ITreeVisitor<TNodeContext> treeV
                         }
 
                         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-                        void VisitMultiThread(TreePath parentPath, ITreeVisitor<TNodeContext> treeVisitor, in TNodeContext nodeContext, ITrieNodeResolver trieNodeResolver, TrieVisitContext visitContext, TrieNode?[] children)
+                        void VisitMultiThread(TreePath parentPath, ITreeVisitor<TNodeContext> treeVisitor, in TNodeContext nodeContext, ITrieNodeResolver trieNodeResolver, TrieVisitContext visitContext)
                         {
-                            var copy = nodeContext;
+                            // we need to preallocate children
+                            TNodeContext contextCopy = nodeContext;
 
-                            // multithreaded route
-                            Parallel.For(0, BranchesCount, RuntimeInformation.ParallelOptionsPhysicalCores, i =>
+                            ArrayPoolList<Task>? tasks = null;
+                            for (int i = 0; i < BranchesCount; i++)
                             {
-                                visitContext.Semaphore.Wait();
-                                try
+                                if (i < BranchesCount - 1 && visitContext.ThreadLimiter.TryTakeSlot(out ThreadLimiter.SlotReturner returner))
                                 {
-                                    TreePath closureParentPath = parentPath;
-                                    // we need to have separate context for each thread as context tracks level and branch child index
-                                    TrieVisitContext childContext = visitContext.Clone();
-                                    VisitChild(ref closureParentPath, i, children[i], trieNodeResolver, treeVisitor, copy, childContext);
+                                    tasks ??= new ArrayPoolList<Task>(BranchesCount);
+                                    tasks.Add(SpawnChildVisit(parentPath, i, GetChild(nodeResolver, ref parentPath, i), returner));
                                 }
-                                finally
+                                else
                                 {
-                                    visitContext.Semaphore.Release();
+                                    VisitChild(ref parentPath, i, GetChild(nodeResolver, ref parentPath, i), trieNodeResolver, treeVisitor, contextCopy, visitContext);
                                 }
-                            });
+                            }
+
+                            if (tasks != null && tasks.Count > 0) Task.WaitAll(tasks.ToArray());
```

Comment:
```suggestion
                            if tasks?.Count > 0) Task.WaitAll(tasks.ToArray());
```

Also we can use do `AsSpan()` instead of `ToArray` in .net 9 (update #7585 when this is merged to master)